packages/api-client-core/src/GadgetConnection.ts

```diff
@@ -491,11 +506,25 @@ export class GadgetConnection {
     });
   }
 
-  private requestHeaders() {
+  private async requestHeaders() {
     const headers: Record<string, string> = {};
 
-    if (this.authenticationMode == AuthenticationMode.InternalAuthToken) {
-      headers.authorization = "Basic " + base64("gadget-internal" + ":" + this.options.authenticationMode!.internalAuthToken!);
+    if (this.authenticationMode == AuthenticationMode.Internal || this.authenticationMode == AuthenticationMode.InternalAuthToken) {
+      const authToken =
+        this.authenticationMode == AuthenticationMode.Internal
+          ? this.options.authenticationMode!.internal!.authToken
+          : this.options.authenticationMode!.internalAuthToken!;
+
+      headers.authorization = "Basic " + base64("gadget-internal" + ":" + authToken);
+
+      if (this.authenticationMode == AuthenticationMode.Internal && this.options.authenticationMode!.internal!.useSession) {
```

Comment:
Are these two options separate because sometimes you want to use a session but you don't know the ID yet or something like that? I feel like it designs into existence passing the getter but not the boolean which is a bit confusing? Can we just use the presence of the get-me-the-id function as the indicator we should be doing this?

Reply:
> Can we just use the presence of the get-me-the-id function as the indicator we should be doing this?

we need both since when you call `api.asSession` we create a new api client, but I want to use a `getSessionId` function that was passed to the original `api` instance

Reply:
this in Client.ts

```
get asSession(): Client {
        assert(this.options?.authenticationMode?.internal, "asSession can only be used for API clients using internal authentication")

        return new Client({
          ...this.options,
          authenticationMode: {
            internal: {
              ...this.options.authenticationMode.internal,
              asSession: true,
            }
          }
        })
      }```